### DISCUSSION TITLE
Autoscaling Llama.cpp server clusters in GPU spot instances.

### DISCUSSION THREAD

**Discussion by jboero** [ggml-org/llama.cpp#6764] [Show and tell]
![llama_autoscale](https://github.com/ggerganov/llama.cpp/assets/7536012/1000b914-af83-46c7-bbcb-602d4fcd8cbb)
I wrote up a Terraform module PoC for autoscaling clusters of Llama.cpp server in GCP spot instances with GPUs. Autoscaling is a bit overzealous but it tends to work pretty well. Anyone curious to try I'd love to hear feedback.

https://github.com/jboero/terraform-google-llama-autoscale

**Comment by phymbert** [comment]
Great, please have a look to:
- #6546 

I believe helm approach is more flexible, but we can also introduce a terraform example.

**Comment by jboero** [comment]
How does the K8s sample work with GPUs though? Each autoscale VM here adds another GPU (or set of GPUs) to the cluster. As I understand most cloud managed Kubernetes options don't support GPUs or customizing kernel/drivers either.

**Reply by phymbert** [reply]
It works really fine with GCP Kubernetes offer, just add a GPU pool, and classical autoscalling works pretty fine.

**Reply by jboero** [reply]
Oh cool that's good to know. Do you know what the cost roughly comes out to be?

**Reply by phymbert** [reply]
Add a GPU pool:

```shell
gcloud container node-pools create xx-gpu \
    --cluster xxx \
    --accelerator type=nvidia-l4,count=2,gpu-driver-version=latest \
    --num-nodes=1 \
    --enable-autoscaling  \
    --min-nodes 0 --max-nodes 2 \
    --zone=europe-west3-b \
    --machine-type g2-standard-24
```

No idea regarding particular pricing override from the machine type pricing.

**Reply by jboero** [reply]
Oh nice this supports spot instances also (`--spot`). It's been a while since I've played with GKE but all checks out at the same rates of $0.08/node for Skylake spot instances in europe-west4. Also it seems the GKE layer doesn't add any management costs. This screenshot from a bare minimal 3 node cluster.

![image](https://github.com/ggerganov/llama.cpp/assets/7536012/b81aa966-01c3-44d2-a105-81fc18a3a6a0)
